features/filesystem/bd/BufferedBlockDevice.h

```diff
@@ -0,0 +1,167 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef MBED_BUFFERED_BLOCK_DEVICE_H
+#define MBED_BUFFERED_BLOCK_DEVICE_H
+
+#include "BlockDevice.h"
+
+
+/** Block device for allowing smaller read and program sizes for the underlying BD,
+ *  using a buffer.
```

Comment:
Not sure if this should be extended but first question I have is probably too implementation detailed? the buffer is static, heap? I can see its on the heap.

Why only smaller ? how much smaller ? Is there a range or any limit?

Reply:
Changed this line a bit. It is actually minimal and not smaller - read and program sizes are fixed to 1.
Reasoning behind such a block device adaptor is the following: Some BDs have read/program sizes that are too big. For instance, the SD card BD has the value of 512 bytes for both. This can make life very hard for users of these BDs, as they will need to add logics for caching data even when they need to read/write small amounts of it. This adaptor solves the problem. One use case (that was the trigger here) was running StoageLite over SD card.